File: 2021/javascript/day_02.js
const data = require('fs').readFileSync('../day_02.data', 'utf-8')

let [h, v] = data.split(/\r?\n/)
  .map(line => line.split(/ /))
  .map(([d, v]) => [d, Number(v)])
  .reduce(([h=0, v=0], [dir, val]) => {
    switch(dir) {
      case "forward": return [h+val, v];
      case "down": return [h, v+val]; 
      case "up": return [h, v-val];
    }
  }, []);

console.log(h * v);

File: 2021/clojure/day_02.clj
(ns day-02)

(def ->int #(Integer/parseInt %))
(def ->cmd (fn [[d v]] [d (->int v)]))

(defn part-one [[h v a] [dir x]]
  (case dir
    "forward" [(+ h x) v a]
    "down" [h (+ v x) a]
    "up" [h (- v x) a]))

(defn part-two [[h v a] [dir x]]
  (case dir
    "forward" [(+ h x) (+ v (* x a)) a]
    "down" [h v (+ a x)]
    "up" [h v (- a x)]))

(defn drive [cmds act]
  (let [[h v] (reduce act [0 0 0] cmds)]
    (* h v)))

(let [lines (re-seq #"[^\n]+" (slurp "../day_02.data"))
      cmds  (map ->cmd (map #(re-seq #"[^ ]+" %) lines))]
  (println "\n-- Part 1 --")
  (println "Horizontal position * depth: " (drive cmds part-one))

  (println "\n-- Part 2 --")
  (println "Horizontal position * depth: " (drive cmds part-two)))